crates/matrix-sdk/src/encryption/recovery/mod.rs

```diff
@@ -0,0 +1,526 @@
+};
+use crate::Client;
+
+pub mod futures;
+mod types;
+pub use self::types::{EnableProgress, RecoveryError, RecoveryState, Result};
+use self::{
+    futures::{Enable, RecoverAndReset, Reset},
+    types::{BackupDisabledContent, SecretStorageDisabledContent},
+};
+
+/// The recovery manager for the [`Client`].
+#[derive(Debug)]
+pub struct Recovery {
+    pub(super) client: Client,
+}
+
+impl Recovery {
+    /// Get the current [`RecoveryState`] for this [`Client`].
+    pub fn state(&self) -> RecoveryState {
+        self.client.inner.recovery_state.get()
+    }
+
+    /// Get a stream of updates to the [`RecoveryState`].
+    ///
+    /// This method will send out the current state as the first update.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// let mut state_stream = recovery.state_stream();
+    ///
+    /// while let Some(update) = state_stream.next().await {
+    ///     match update {
+    ///         RecoveryState::Enabled => {
+    ///             println!("Recovery has been enabled");
+    ///         }
+    ///         _ => (),
+    ///     }
+    /// }
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub fn state_stream(&self) -> impl Stream<Item = RecoveryState> {
+        self.client.inner.recovery_state.subscribe_reset()
+    }
+
+    /// Enable secret storage *and* backups.
+    ///
+    /// This method will create a new secret storage key and a new backup if one
+    /// doesn't already exist. It will then upload all the locally cached
+    /// secrets, including the backup recovery key, to the new secret store.
+    ///
+    /// This method will throw an error if a backup already exists on the
+    /// homeserver but this [`Client`] isn't connected to the existing backup.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::EnableProgress};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// let enable = recovery
+    ///     .enable()
+    ///     .wait_for_backups_to_upload()
+    ///     .with_passphrase("my passphrase");
+    ///
+    /// let mut progress_stream = enable.subscribe_to_progress();
+    ///
+    /// tokio::spawn(async move {
+    ///     while let Some(update) = progress_stream.next().await {
+    ///         let Ok(update) = update else {
+    ///             panic!("Update to the enable progress lagged")
+    ///         };
```

Comment:
Does a simple unwrap not work?

Reply:
Well I wanted to put a sensible message into the example and using `expect()` with the message produces an overly long line.